### PR TITLE
[RLlib] Update max_seq_len in pad_batch_to_sequences_of_same_size

### LINKED ISSUE #19976: [Bug] [rllib] RNN sequencing is incorrect
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I would expect given two sequences `A, B`:
`[A, A, A, B, B]; seq_lens=[3, 2], obs.shape = [5, 1]`
would be padded to
`[A, A, A, B, B, *]; seq_lens=[3, 2], obs.shape = [2, 3, 1]`

This does not appear to be the case. For some reason rllib zero-pads `obs` to something besides `seq_lens.max()`. Even more worrisome is calling `torch.nonzero()` on the `input_dict`, which shows front-padded zeros to the observations. For example, printing `input_dict['obs'].reshape(B, T, -1) == 0` results in:

```
(PPO pid=74137)         [[ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True]],
```

The zero-padding is clearly messed up, the first five observations have been zero-padded and then we have real observations offset by five.

### Versions / Dependencies

Linux Ray 1.7.0

### Reproduction script
Feel free to play with the `USE_CORRECT_SHAPE` flag

```python
import torch
import numpy as np
import gym
from typing import Union, Dict, List, Tuple, Any
import ray
from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
from ray.rllib.utils.typing import ModelConfigDict, TensorType
from ray.rllib.policy.rnn_sequencing import add_time_dimension
from ray.tune import register_env
from ray.rllib.agents.ppo import PPOTrainer
from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole

# Pad to the correct size and crash
# or follow the rnn_sequencing code and don't crash
USE_CORRECT_SHAPE = False

class TestRNN(TorchModelV2, torch.nn.Module):
    def __init__(
        self,
        obs_space: gym.spaces.Space,
        action_space: gym.spaces.Space,
        num_outputs: int,
        model_config: ModelConfigDict,
        name: str,
        **custom_model_kwargs,
    ):
        TorchModelV2.__init__(
            self, obs_space, action_space, num_outputs, model_config, name
        )
        torch.nn.Module.__init__(self)
        self.num_outputs = num_outputs
        self.input_dim = gym.spaces.utils.flatdim(obs_space)
        self.act_space = action_space
        self.act_dim = gym.spaces.utils.flatdim(action_space)
        self.cur_val = None

        self.policy = torch.nn.Linear(self.input_dim, self.act_dim)
        self.vf = torch.nn.Linear(self.input_dim, 1)

    def get_initial_state(self):
        return [torch.zeros(0)]

    def value_function(self):
        assert self.cur_val is not None, "must call forward() first"
        return self.cur_val

    def forward(
        self,
        input_dict: Dict[str, TensorType],
        state: List[TensorType],
        seq_lens: TensorType,
    ) -> Tuple[TensorType, List[TensorType]]:

        flat = input_dict["obs_flat"]

        if USE_CORRECT_SHAPE:
            max_seq_len = seq_lens.max()
        else:
            # max_seq_len here is copied from rllib RNN code
            # see https://github.com/ray-project/ray/blob/2d24ef0d3234867ac329b10ae3a11b9b7119d17b/rllib/models/torch/recurrent_net.py#L75
            # but it doesn't make sense...
            # it should be max_seq_len = seq_len.max()
            max_seq_len = flat.shape[0] // seq_lens.shape[0]

        padded = add_time_dimension(
            flat,
            max_seq_len=max_seq_len,
            framework="torch",
            time_major=False
        )

        B = padded.shape[0]
        T = padded.shape[1]

        # If this fails, then we have "extra" padding in the RNN
        # We shouldn't need to pad the time dimension more than the longest
        # sequence
        if seq_lens.max() != T:
            print(f'seq_lens.max() is {seq_lens.max()} but input temporal dim is {T}')
            print(flat.reshape(B, T, -1) == 0)
            raise Exception('seq_len mismatch')

        flattened = padded.reshape(-1, padded.shape[-1])
        logits = self.policy(flattened)
        self.cur_val = self.vf(flattened).squeeze(1)
        state = state

        return logits, state

register_env(StatelessCartPole.__name__, StatelessCartPole)
MAX_SEQ_LEN = 200
CFG = {
    "env_config": {},
    "framework": "torch",
    "model": {
        "custom_model": TestRNN,
        "max_seq_len": MAX_SEQ_LEN,
    },
    "num_workers": 0,
    "num_gpus": 0,
    "env": StatelessCartPole,
    "horizon": MAX_SEQ_LEN,
}
ray.init(object_store_memory=3e10)
analysis = ray.tune.run(
    PPOTrainer,
    config=CFG,
)
```

### Anything else

Every train step

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by mvindiola1**
## Why are these changes needed?
There where two issues causing the associated issue.

1. pad_batch_to_sequences_of_same_size was not updating max_seq_len when dynamic_max=True. This caused sequence slice alignment issues when doing sgd in ppo with an RNN or Attention model

2. When using an RNN or Attention model the sgd_minibatch_size must be >= to max_seq_len otherwise it will cause slice index issues n doing sgd in ppo

I also added a tests max_seq_len in pad_batch_to_sequences_of_same_size for when dynamic_max is True or False.

## Related issue number
Closes #19976 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [N/A] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mvindiola1** [comment]
Well that failed. There are several rllib tests that set the mini-batch size to 1 or 10. The new check in validate_config raises an exception because they are less than max_seq_len.

I could not figure out a reliable way to check for states in validate config. use_lstm/attention would not work in the case of a custom model.

I suppose instead of a value error we could print a warning or we could change max_seq_len I. These tests.

**Comment by mvindiola1** [comment]
cc @sven1977 @gjoliver

File: rllib/agents/ppo/ddppo.py [hunk 1]
```diff
@@ -129,6 +129,15 @@ def validate_config(self, config):
                 "Set rollout_fragment_length instead of train_batch_size "
                 "for DDPPO.")
 
+        # Check to make sure that the sgd_minibatch_size is not smaller
+        # than max_seq_len otherwise this will cause indexing errors while
+        # performing sgd when using a RNN or Attention model
+        if config["sgd_minibatch_size"] < config["model"]["max_seq_len"]:
```
Comment (gjoliver) [inline rllib/agents/ppo/ddppo.py:135]: thinking about this a bit more, will this work for non-rnn stacks? there are a few tests failing because of this new check.
still feels a bit weird that sgd_minibatch_size is connected to max_seq_len now.
I wonder if this requirement should just be a warning or something.

Comment (mvindiola1) [reply rllib/agents/ppo/ddppo.py:135]: @gjoliver,

I think the question is what do you want to happen if a user sets max_seq_len > sgd_minibatch_size

The current code throws an exception but it only does so during training and only after the policy learns to generate episodes that last longer than sgd_minibatch_size. In StatelessCartpole it has to train for ~50 iterations before it asserts. That is a long time to wait to find out your config is set up incorrectly. The current assertion does not tell you that the config is the issue so you have to decode that.

One way to fix this is to slice based on min of (max_seq_len, sgd_minibatch_size) but if you do this you are ignoring the users request that truncated-bptt should be at most max_seq_len. So in the example in the issue you would be changing 200 -> =128. 


A second way to fix this is to test the configuration and warn the user ahead of time. The question is where to test. 

a. validate_config: Easy to test the config but no easy way to see if the model uses state and max_seq_len should even be considered. Currently some tests are failing because they fail the check but it should not matter because they do not use states.

b. At policy construction time. Then we could check policy.is_recurrent() and only if it is True do we do this config check.

c. Put it in do_minibatch_sgd. This might be the real place we want to put it since it is not actually about PPO but the sgd process.

Comment (mvindiola1) [reply rllib/agents/ppo/ddppo.py]: I just pushed an update that tries to implement option c.

**Review by gjoliver — APPROVED** [review]
ok, this only applies to recurrent network. that's much safer.
this seems very reasonable to me.

@sven1977, can you help merge please?
thanks.
